max30102.c

```diff
@@ -157,65 +267,165 @@ MAX30102_Status_t OFF(I2C_HandleTypeDef *hi2c) {
+    static uint8_t lastTenBeatCounts[10];
+    static uint32_t lastIRValue = 0;
+
+    for (uint8_t i = 0; i < unreadSamples; i++) {
+        if (tempBuffer[i].iRed < 40000) {
+            *heartRate = 0;
+            *spo2 = 0;
+            eachSampleDiff = 0;
+            continue;
         }
 
-        // Get current time in milliseconds, function exists in all STM32 boards
-        curr_time = HAL_GetTick();
+        MAX30102_InsertSample(tempBuffer[i]);
+        MAX30102_CalculateAcDc(&redAC, &redDC, &irAC, &irDC);
+        MAX30102_FilterSample(&tempBuffer[i]);
 
-        // Check and update peak value for the max infrared value, only update when an upward trend in infrared data value occurs.
-        if (ir_data > prev_ir_data && ir_data > peak_data_signal) { // Double checks to ensure peak value not updated on random highs
-            peak_data_signal = ir_data;  // Update peak if current value is higher
-        } 
+        float ratio = ((float)redAC / redDC) / ((float)irAC / irDC);
+        if (ratio >= 0.36 && ratio < 0.66) {
+            *spo2 = (uint8_t)(107 - 20 * ratio);  // Use *spo2 to dereference
+        } else if (ratio >= 0.66 && ratio < 1) {
+            *spo2 = (uint8_t)(129.64 - 54 * ratio);  // Use *spo2 to dereference
+        }
+
+        eachSampleDiff = lastIRValue - tempBuffer[i].iRed;
+        if (eachSampleDiff > 50 && beatSampleCount > 12) {
+            for (uint8_t j = 9; j > 0; j--) {
+                lastTenBeatCounts[j] = lastTenBeatCounts[j - 1];
+            }
+            lastTenBeatCounts[0] = beatSampleCount;
 
-        // Peak value is found, calculates time interval between new peak value and previous peak value
-        else if (ir_data < prev_ir_data && peak_data_signal > 0) {
-            if (peaks_reached > 0) {
-                peak_interval_sum += curr_time - prev_peak_time;
+            uint32_t totalTime = 0;
+            for (uint8_t j = 0; j < 10; j++) {
+                totalTime += lastTenBeatCounts[j];
             }
 
-            prev_peak_time = curr_time;  
-            peak_data_signal = 0;
-            peaks_reached++; 
+            *heartRate = (uint8_t)(60.0 * 10 / 0.02 / totalTime);  // Use *heartRate to dereference
+            beatSampleCount = 0;
         }
 
-        prev_ir_data = ir_data; // Assign current ir data value the previous one. 
+        lastIRValue = tempBuffer[i].iRed;
+        beatSampleCount++;
+    }
 
-        // Calculate heart rate in BPM after every 15 peaks. More peaks = more accurate measurements since the average heart rate is calculated with more interval values
-        if (peaks_reached >= num_peaks) {
-            *heart_rate = (60 * 1000) / (peak_interval_sum / peaks_reached); // Convert average interval time from milliseconds to minutes 
-            peak_interval_sum = 0; 
-            peaks_reached = 0;         
-        }
-    return MAX30102_OK; // Return if loop is working properly
+    return MAX30102_OK;
+}
```

Comment:
I got the study and inspiration for the logic from [https://github.com/julpage](https://github.com/julpage)